dbms/src/Common/getNumberOfLogicalCPUCores.cpp

```diff
@@ -0,0 +1,210 @@
+    // The CPU numbers are comma-separated numbers or ranges. For example:
+
+    // # cat cpuset.cpus
+    // 0-4,6,8-10
+
+    // An empty value indicates that the cgroup is using the same setting as the nearest cgroup ancestor with a non-empty "cpuset.cpus" or all the available CPUs if none is found.
+    // The value of "cpuset.cpus" stays constant until the next update and won't be affected by any CPU hotplug events.
+    std::ifstream infile(filename);
+    if (!infile.is_open())
+    {
+        return default_value;
+    }
+    std::string line;
+    std::getline(infile, line);
+    int cpu_count = 0;
+    for (std::string::size_type first = 0, i = 0; i != std::string::npos; i = line.find(','))
+    {
+        std::string cpu_set = line.substr(first, i);
+        if (cpu_set.find('-') != std::string::npos)
+        {
+            std::string start_str = cpu_set.substr(0, cpu_set.find('-'));
+            std::string end_str = cpu_set.substr(cpu_set.find('-') + 1);
+            int start = std::stoi(start_str);
+            int end = std::stoi(end_str);
+            cpu_count += end - start + 1;
+        }
+        else
+        {
+            cpu_count++;
+        }
+        first = i + 1;
+    }
+    return cpu_count;
+}
+
+static std::pair<int, int> read_quota_and_period_from(const char * filename)
+{
+    // cpu.max
+    // A read-write two value file which exists on non-root cgroups. The default is "max 100000".
+    // The maximum bandwidth limit. It's in the following format:
+
+    // $MAX $PERIOD
+
+    // which indicates that the group may consume upto $MAX in each $PERIOD duration.
+    // "max" for $MAX indicates no limit. If only one number is written, $MAX is updated.
+    std::ifstream infile(filename);
+    if (!infile.is_open())
+    {
+        return std::make_pair(1, 1);
+    }
+    std::string quota;
+    int period;
+    infile >> quota >> period;
+    if (quota == "max")
+    {
+        return std::make_pair(period, period);
+    }
+    else
+    {
+        return std::make_pair(std::stoi(quota), period);
+    }
+}
+
+unsigned calCPUCores(int cgroup_quota, int cgroup_period, int cgroup_share, unsigned default_cpu_count)
+{
+    unsigned quota_count = default_cpu_count;
+
+    if (cgroup_quota > -1 && cgroup_period > 0)
+    {
+        quota_count = ceil(static_cast<float>(cgroup_quota) / static_cast<float>(cgroup_period));
+    }
+
+    // Convert 1024 to no shares setup
+    if (cgroup_share == 1024)
+        cgroup_share = -1;
+
+#define PER_CPU_SHARES 1024
+    unsigned share_count = default_cpu_count;
+    if (cgroup_share > -1)
+    {
+        share_count = ceil(static_cast<float>(cgroup_share) / static_cast<float>(PER_CPU_SHARES));
+    }
+
+    return std::min(default_cpu_count, std::min(share_count, quota_count));
+}
```

Comment:
https://github.com/ClickHouse/ClickHouse/pull/35815

Checkout whether we should adapt to these changes pls

Reply:
I think it is reasonable, `cpu.shares` only works when the machine is free, so cpus should be visible to tiflash. I will adapt it soon.